[MM-53432] Calls transcriptions

streamer45 opened the pull request:
#### Summary

PR implements the plugin side for calls transcriptions (post call transcript). 

One important detail to keep in mind is that recording and transcription jobs are artificially coupled, meaning that when transcriptions are enabled (globally from the plugin settings), starting and stopping recordings will automatically start and stop transcriptions. Moreover, any failure would cause both jobs to be terminated.

One limitation I found when trying to follow the proposed design is that it's not really possible to update the attachments for a previously created post so we cannot easily add the transcription file to recording post unless we artificially delay its creation. I don't like that idea very much given the transcription could take a while to process. For now we simply have the bot make a new post with the transcription file attached. I created https://mattermost.atlassian.net/browse/MM-54874 to track this improvement as it'll likely require server changes. 

**Still to do** (ideally prior to merging)

- [ ] Fix E2E pipeline. Need to create and bump the builder to use Go 1.21. (https://git.internal.mattermost.com/qa/calls-e2e-testing/-/merge_requests/37)
- [ ] Need to prepare a test server for UX/PM to test this on.
- [ ] At least a couple of e2e tests. Need to update all the pipelines to make it happen though.
- [ ] Attaching a text based format in addition to WebVTT for easy (human readable) preview of the transcription content.
- [ ] Probably need to slightly rework the transcript format since the speaker names are not rendered automatically.
- [ ] Add some basic configuration for transcription jobs, such as quality (e.g. model size to use).
- [ ] Consider having a setting to control the timeout for the transcribing job given we currently rely on `MaxRecordingDuration` (see comment in code).
- [ ] Update documentation (e.g. the new plugin setting).
- [ ] Merge, release and update all the dependencies (5 sister repositories at least I believe).

#### Design

https://www.figma.com/file/ZAvwHhdUTaWkby4uekmnDX/Call-transcription

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-53432

cpoile commented on server/session.go:
```diff
@@ -320,22 +328,66 @@ func (p *Plugin) removeSession(us *session) error {
-		p.LogDebug("all users left call with recording in progress, stopping", "channelID", us.channelID, "jobID", currState.Call.Recording.JobID)
-		if err := p.getJobService().StopJob(us.channelID); err != nil {
-			p.LogError("failed to stop recording job", "error", err.Error(), "channelID", us.channelID, "jobID", currState.Call.Recording.JobID)
+	// Checking if the transcription has ended due to the bot leaving.
+	if prevState.Call != nil && prevState.Call.Transcription != nil && currState.Call != nil && currState.Call.Transcription != nil &&
+		currState.Call.Transcription.EndAt > prevState.Call.Transcription.EndAt {
```
I'm starting to think the way that the bot behaviour + state is triggering things seems a little brittle and confusing.  Maybe it's too late to change it, but it would be nice if jobs were started and stopped by idempotent events. Granted, we need the bots to do any of the actual recording/transcribing, but it seems like maybe the communication triggers and the stopping triggers (at least) could be in response to events, rather than these  conditional blocks (which seems like it would be error prone).

streamer45 replied:
I'd like to understand more about that as I am not entirely sure what you mean by "idempotent events". Could you make an example perhaps?

In essence the code above is in response to an event, namely the bot leaving the call which we need to handle somehow. Not sure if it's a syntactic improvement you'd be looking for (e.g. `onBotLeftCall` callback) or something more structural.

cpoile replied:
Here's my thinking (and while writing this I think I see why it's not practical): Something like moving all the "bot leaving triggers the recording/transcription ending" logic out of here and putting it in its own function (say: `recordingEnded`), then moving the conditional logic (lines 350-351) somewhere else (that's the infeasible part I think), then using that conditional logic to trigger an api call which trigger the `recordingEnded`. Then if that `recordingEnded` endpoint gets triggered multiple times (for whatever reason) we can ignore it after the first one. 
I think it's tough to find a better place for lines 350-351 though...
I guess I'm worried that we've tightly coupled the bot's presence with the recording/transcribing activity. I know we need the bot there for that to actually work, I just feel it's mixing two concepts that logically should be separate...

streamer45 replied:
Alright, I feel this deserves a sync discussion, let's chat about it next week.

cpoile commented on server/transcription_api.go:
```diff
@@ -0,0 +1,192 @@
+		clientState.Err = "failed to start transcriber job: timed out waiting for bot to join call"
+		clientState.EndAt = time.Now().UnixMilli()
+
+		if state.Call.Recording != nil && state.Call.Recording.EndAt == 0 {
+			if _, _, err := p.stopRecordingJob(state, callID); err != nil {
+				p.LogError("failed to stop recording job", "callID", callID, "err", err.Error())
+			}
+		}
```
My understanding: we're stopping the recording if the transcriber bot fails to join, is that correct? I guess it keeps them linked... Maybe that is the best thing to do. (leaving this comment here just in case it sparks any ideas)

cpoile commented on server/transcription_api.go:
```diff
@@ -0,0 +1,192 @@
+		return fmt.Errorf("failed to create transcription job: %w", jobErr)
+	}
+
+	if trState.JobID != "" {
+		return fmt.Errorf("transcription job already in progress")
+	}
```
That's funny. Not sure how that would happen :)

streamer45 replied:
It actually can happen because we are not 100% atomic there due to the fact that we need to unlock to make the API call to the job service. So concurrent requests to start transcriptions could theoretically result in multiple jobs starting. Of course the client shouldn't ever do that but it's a possibility we may want to account for in case of issues/bugs.

cpoile replied:
Ah, good to know, thanks

cpoile commented on server/transcription_api.go:
```diff
@@ -0,0 +1,192 @@
+	return nil
+}
+
+func (p *Plugin) stopTranscribingJob(state *channelState, callID string) (rerr error) {
```
Is there an opportunity to consolidate some code with the checker? Not a big deal though.

cpoile reviewed (commented):
Just minor comments, nothing big. Amazing work!

antran22 commented:
This is awesome, big thanks to Mattermost Team for this feature. Is there any point on the roadmap for this feature that we can wait for?

streamer45 commented on e2e/config.ts:
```diff
@@ -0,0 +1,40 @@
+};
+
+export const apiPatchConfig = async (cfg: CallsConfig) => {
+    const adminContext = (await newUserPage(adminState.storageStatePath)).page.request;
```
We need an admin context so it's a bit more tricky as we can't just re-use the existing one from the test. Probably setting up a request is the way to go here. I'll look into it.

streamer45 replied:
Done in 32f26ac

streamer45 commented on plugin.json:
```diff
@@ -179,6 +179,13 @@
+        "display_name": "Enable call transcriptions (Experimental)",
+        "type": "bool",
+        "default": false,
+        "help_text": "(Optional) When set to true it enables the call transcriptions functionality."
```
I'd probably keep it as is for consistency with recordings. What the functionality entails should probably be explained by documentation.

streamer45 commented on server/bot_api.go:
```diff
@@ -207,43 +218,153 @@ func (p *Plugin) handleBotPostRecordings(w http.ResponseWriter, r *http.Request,
+		return
+	}
+
+	postID := info["thread_id"]
```
I think the ideal solution would be for the request body to be properly typed and likely provide an `IsValid()` method. That means more coupling as both sides need to be aware of the type but I think we are going into that direction anyway with the new `public` stuff so may as well do it.

cpoile replied:
(again, just an idea, not necessary): That's why a more generic `ensureFields(stringmap map[string]string, fields ...string)` type of thing would be better? (less coupling)

streamer45 replied:
Yep, understood the proposal. My main worry is that that would become problematic as soon as we need to handle something other than strings. And then we are re-inventing structs using maps. Maybe some coupling is the way to go given this is essentially a public API from the implementer's point of view.

streamer45 replied:
So I added a `JobInfo` type and updated everything, no more map parsing on the API side.

streamer45 commented on server/bot_api.go:
```diff
@@ -273,61 +394,126 @@ func (p *Plugin) handleBotPostJobsStatus(w http.ResponseWriter, r *http.Request,
+
+	state, err := p.lockCall(callID)
+	if err != nil {
+		p.LogError("handleBotGetProfileForSession: failed to lock call", "err", err.Error())
+		res.Code = http.StatusInternalServerError
+		res.Err = err.Error()
 		return
 	}
+	defer p.unlockCall(callID)
 
-	res.Err = "bad request"
-	res.Code = http.StatusBadRequest
+	if state == nil || state.Call == nil {
+		res.Code = http.StatusBadRequest
+		res.Err = "no call ongoing"
+		return
+	}
```
What do you mean by "move the check into the lock"? The whole function is under lock.

cpoile replied:
I mean the "lock and check if there is a call and return if there isn't" flow. If we include the check in the lock we can maybe simplify? Not a big deal if it's not worth it, it's just a thought.

streamer45 replied:
Oh do you mean to include the call existence check directly in `lockCall()`? I think that would be a problem as we also lock to start a call in which case it's expected for the call state not to exist.

streamer45 commented on server/job_service.go:
```diff
@@ -203,16 +215,32 @@ func (p *Plugin) getJobService() *jobService {
 }
 
-func (s *jobService) StopJob(channelID string) error {
+func (s *jobService) StopJob(channelID, connID string) error {
```
Yep, that should be the case.

streamer45 replied:
Fixed in 95ed7141ac3bab35b361586448e0cad993770c20

streamer45 commented on server/job_service.go:
```diff
@@ -314,14 +358,20 @@ func (p *Plugin) initJobService() error {
 	}
 
-	if err := jobService.Init(recordingJobRunner); err != nil {
+	if err := jobService.Init([]string{recorderJobRunner, transcriberJobRunner}); err != nil {
```
Yeah but we are just passing the runner so that the docker image can be prefetched if needed. I think that's fine as transcriptions could be turned on at a later time without having to restart everything. But not super opinionated either.

cpoile replied:
I just wonder if people will be confused by it prefetching the transcriber image (looking like it's getting ready for it) but they haven't enabled it? (just thinking, not necessary)

streamer45 replied:
Sure, I can get on board with the "don't do it if not needed" reasoning.

streamer45 commented on server/recording_api.go:
```diff
@@ -60,32 +68,26 @@ func (p *Plugin) recJobTimeoutChecker(callID, jobID string) {
-func (p *Plugin) handleRecordingStartAction(state *channelState, callID, userID string) (*RecordingStateClient, httpResponse) {
-	var res httpResponse
-
+func (p *Plugin) startRecordingJob(state *channelState, callID, userID string) (rst *JobStateClient, rcode int, rerr error) {
```
The reason to have named returns is so that we can intercept errors in the defer scope. But Go doesn't allow to have a single named return, it's either all of them or none, hence the change.

cpoile replied:
Ahh, I missed that connection, thanks.

streamer45 commented on server/metadata.go:
```diff
@@ -0,0 +1,121 @@
+			RecID: recID,
+		}
+
+		transcriptions, ok := post.GetProp("transcriptions").(map[string]any)
+		if !ok {
+			transcriptions = map[string]any{
+				trID: tm.toMap(),
+			}
+		} else {
+			recordings[trID] = tm.toMap()
+		}
+		post.AddProp("transcriptions", transcriptions)
```
Can't tell if you are being sarcastic :p

cpoile replied:
Definitely sarcastic. :)

streamer45 commented on server/metadata.go:
```diff
@@ -0,0 +1,121 @@
+	FileID string
+	// PostID is the Post.Id that holds the job's artifacts (e.g. recording post).
+	PostID string
+	// RecID is the recording ID.
+	RecID string
+	// TrID is the transcription ID.
+	TrID string
```
I think we lose some clarity if we do that because in the recording job's metadata we store the job id for the related transcription and vice versa. If we simply call it `JobID` I feel it gets even more confusing. Plus we already omit empty fields so there's isn't a huge concern.

cpoile replied:
That makes sense.

streamer45 commented on e2e/tests/recordings.spec.ts:
```diff
@@ -52,5 +57,52 @@ test.describe('call recordings', () => {
+        await page.locator('.ThreadViewer').locator('.post__body').nth(1).filter({has: page.getByTestId('fileAttachmentList')}).click();
+        await expect(page.locator('.file-preview-modal__content')).toBeVisible();
+
+        // verify transcription track exists
+        await expect(page.getByTestId('calls-recording-transcription')).toHaveAttribute('label', 'Transcription');
```
Done in a9e10dc

streamer45 commented:
> This is awesome, big thanks to Mattermost Team for this feature. Is there any point on the roadmap for this feature that we can wait for?

Hopefully sometime next month. But the functionality (at least initially) will only be available for Enterprise licensed installations.

streamer45 commented on server/bot_api.go:
```diff
@@ -273,61 +394,126 @@ func (p *Plugin) handleBotPostJobsStatus(w http.ResponseWriter, r *http.Request,
-			return
-		}
+	if jb.ID != jobID {
+		res.Err = "invalid recording job ID"
```
Fixed in 49c5670482b5559de05e7faa7ec8a47eeb060a9e

streamer45 commented:
@cpoile Asking for another pass, mostly to move the conversations forward. I still have to address a couple of things.

streamer45 commented on webapp/src/components/recordings_file_preview.tsx:
```diff
@@ -0,0 +1,52 @@
+}
+
+const RecordingsFilePreview = ({fileInfo, post}: Props) => {
+    const callPost = useSelector((state: GlobalState) => getPost(state, post.props?.call_post_id));
+
+    const recording = callPost?.props?.recordings?.[post.props?.recording_id];
+    const transcription = callPost?.props?.transcriptions?.[recording?.tr_id];
```
Yes I suppose that could be possible. We didn't have the bot when we originally designed the call thread. I suppose we could revisit and have the bot create the post, or maybe keep it this way but disallow user edits through the `MessageWillBeUpdated` plugin hook. I'll create a ticket so we don't forget about this.

streamer45 commented:
@cpoile Heads up I am going to have to change some e2e tests checks slightly as we can't guarantee whether we'll be just recording or also transcribing since these tests will run concurrently with the transcription test.

https://github.com/mattermost/mattermost-plugin-calls/blob/7d981585bacadbc9e5eebb369d4dfe7ba8155e98/e2e/tests/popout.spec.ts#L162-L164

cpoile reviewed (approved):
Looks nice, thank you!

matthewbirtch reviewed (commented):
This is really awesome work @streamer45. 

1. Can we add an obvious primary `Dismiss` button to this notice:
<img width="506" alt="image" src="https://github.com/mattermost/mattermost-plugin-calls/assets/2040554/99680e2c-8705-4d43-9a44-a685762ba441">

So that it looks more like this:
![image](https://github.com/mattermost/mattermost-plugin-calls/assets/2040554/18010f5b-6b32-41d7-bb46-dc005bcd321c)

2. For the filename, my apologies, I don't think I did a very good job providing direction on that. Can we change the format to: `Call_Channel-Name_YYYY-MM-DD_HH-MM-SS` - for example:
    - Channel: `Call_UX-Design_2023-11-03_11-33-22.vtt`
    - DM: `Call_Matthew-Birtch_2023-11-03_11-33-22.vtt`
    - GM: `Call_Claudio-Matthew-Chris_2023-11-03_11-33-22.vtt`

For the group message name above, we'll probably have to truncate the name to cut it off at 25 characters or something.

3. As per Eric's comment in the thread, were you able to tune things at all to get better splitting of text? Even though I was speaking with a constant flow, it was breaking it up in to 5-word chunks. Or is that necessary in order to display captions properly?

4. In the transcript txt file, can we add a `:` after the speaker's name? This might help better separate from the spoken words.

Nothing really blocking here, but let me know if you have any questions.

streamer45 commented:
Thanks @matthewbirtch

1. Yep, will do.
2. I only added the channel name if open/private as I wasn't sure how to deal with DMs/GMs. Will update that and tweak the time format as provided.
3. I haven't looked into words splitting just yet. And funnily enough in my tests I am consistently getting very long sentences. I'll do some tests and report back.
4. Yes, we have full control on that. To confirm, do you mean in the attached file or the video rendered format? Because we already have a new line in the attached plaintext file. To clarify:

Attached text format:

```
00:00:02 -> 00:00:08
Claudio
Testing transcriptions
```

Video rendered format:

`(Claudio) Testing transcriptions`

matthewbirtch commented:
> Had a look and we are not limiting the size of the segments in any way so whatever split it's doing I think it's just up to the model. We can make sentences shorter but not longer it seems. Would you be able to send me a sample recording that shows this problem? I am really having a hard time to reproduce it unless I pause a lot when talking.

No problem, here is the recording and transcript.

[call-x5iazqq56ifaxnfkiyyq4e7xha-2023-11-10-21_14_19.txt](https://github.com/mattermost/mattermost-plugin-calls/files/13324170/call-x5iazqq56ifaxnfkiyyq4e7xha-2023-11-10-21_14_19.txt)

https://github.com/mattermost/mattermost-plugin-calls/assets/2040554/354eff33-70b2-4b86-8277-409bbed8b81c

streamer45 commented:
Thanks @matthewbirtch. By forcing a single text segment I was able to get it to output extremely long sentences. So we are close to two extremes. According to most recommendations I could find, subtitles lines should be kept at around ~40 characters maximum length. I suppose we could try to add some heuristic to stitch contiguous text segments together if the following conditions apply:

- There are no substantial gaps (e.g. a few seconds) between the segments.
- Segments length combined is not exceeding the maximum size in characters.

That way we may be able to get a slightly more compact output.

streamer45 commented:
Thinking more on the above I realize we don't necessarily have to output the same text segments in both formats so we could keep the captions as they are but make the attached `txt` file more compact (longer sentences). 

By the way, I updated the server with some improvements, namely:

- Added dismiss button to the banner.
- Filenames should now follow the suggested format based on the channel type.
- Enabled multi-threading support so the process should take a little less to produce a transcription now.
- Added multilingual models support. Language to use is now auto detected based on the speech found.

Still waiting for the above discussion to unfold to make further improvements.

 /cc @matthewbirtch @esethna

matthewbirtch commented:
> Thinking more on the above I realize we don't necessarily have to output the same text segments in both formats so we could keep the captions as they are but make the attached `txt` file more compact (longer sentences).

Love this idea. I think that makes a tonne of sense. Would time stamps still be represented properly in the txt file? The reason I ask is if we decide to do something more fancy like zoom does for the transcription with the video player ([as Abhijit designed](https://www.figma.com/file/ZAvwHhdUTaWkby4uekmnDX/Call-transcription?type=design&node-id=1002-121547&mode=design&t=6bywEijubuf93y31-4)), would this prevent us from doing this? If not, then I say let's go for this.

> By the way, I updated the server with some improvements, namely:
> 
> * Added dismiss button to the banner.
> * Filenames should now follow the suggested format based on the channel type.
> * Enabled multi-threading support so the process should take a little less to produce a transcription now.
> * Added multilingual models support. Language to use is now auto detected based on the speech found.

Awesome, these are really great updates. And the transcription processing time is WAY faster now.

streamer45 commented:
> Love this idea. I think that makes a tonne of sense. Would time stamps still be represented properly in the txt file? The reason I ask is if we decide to do something more fancy like zoom does for the transcription with the video player ([as Abhijit designed](https://www.figma.com/file/ZAvwHhdUTaWkby4uekmnDX/Call-transcription?type=design&node-id=1002-121547&mode=design&t=6bywEijubuf93y31-4)), would this prevent us from doing this? If not, then I say let's go for this.

To implement that design we'd likely be reading from the existing VTT file anyway since it will contain more precise timestamps. We may even have to include extra metadata or work on a custom format to make it work as we may need to store user ids (as opposed to display names) if we want the final component to feel just like a thread view.

But regardless I think we can pretty much decouple how we format the attached text file as it's purely meant for user ingestion so we can customize as needed.

streamer45 commented:
@matthewbirtch @esethna Is there anything else blocking you were looking to see added/changed? I am still working on some small improvements on the text output as discussed above but other than that I think we are ready to move forward. Once approved I can begin the process of getting this in Community, ideally sometime this week.

@cwarnermm Would be great if you could give this a pass when you get a chance. We are introducing some strings in the following files:

https://github.com/mattermost/mattermost-plugin-calls/pull/549/files#diff-7b32f67ab72ef7b295c60be75ff703b7660e0f7e737e77251e5cc69deb7844aa

https://github.com/mattermost/mattermost-plugin-calls/pull/549/files#diff-bd9c8eded3f2f6d40191d91defa64d238b56387d4f61e3bdb75140d02cb1bb80

matthewbirtch commented:
>  Is there anything else blocking you were looking to see added/changed? I am still working on some small improvements on the text output as discussed above but other than that I think we are ready to move forward. Once approved I can begin the process of getting this in Community, ideally sometime this week.

Nope, nothing else for me other than the text output. I think we're good to proceed as well @streamer45. Great work! I'll approve this knowing we will be doing some tweaking to the text output and we can test things out more robustly on Community.

streamer45 commented:
@matthewbirtch I implemented some basic heuristics to compact the produced text file output in order to avoid very short sentences. I tried a couple of approaches and eventually converged on a duration based one (as opposed to length) since I felt that splitting/joining based on time gives better results overall as it accounts for pace and potential pauses in speech. 

It's still experimental and highly configurable, right now enforcing the following rules. We join segments if:

1. The speaker doesn't change. This is required to guarantee order of the segments (e.g. question/answer sequences).
2. There are less than X seconds of pause between the end of a previous text segment and the start of the next one. (X = 2s)
3. The overall duration of the current sentence is less than Y seconds. (Y = 10s)

Give it a try when you get a chance and let me know if you see any improvements or hit any issue.

cwarnermm commented on plugin.json:
```diff
@@ -179,6 +179,35 @@
+        "display_name": "Enable call transcriptions (Experimental)",
+        "type": "bool",
+        "default": false,
+        "help_text": "(Optional) When set to true it enables the call transcriptions functionality."
```
```suggestion
        "help_text": "(Optional) When set to true, call transcriptions are enabled."
```

cwarnermm commented on webapp/i18n/en.json:
```diff
@@ -1,12 +1,12 @@
   "0UG4R4": "Connecting to the call…",
   "0cE6s2": "Calls can't be initiated in an insecure context",
   "0fFRUQ": "Start presenting",
+  "1Gs1ii": "You can find the recording and transcription in this call's chat thread once it's finished processing.",
```
```suggestion
  "1Gs1ii": "You can find the recording and transcription in this call's chat thread once it has finished processing.",
```

cwarnermm commented on webapp/i18n/en.json:
```diff
@@ -143,6 +149,7 @@
   "n1Ubod": "Enable call recordings",
   "nOZzE1": "You're viewing {presenter}'s screen",
   "neS8Zu": "You're viewing {presenterName}'s screen",
+  "nlQg/n": "The host has started recording and transcription for this meeting. By staying in the meeting you give consent to being recorded and transcribed.",
```
```suggestion
  "nlQg/n": "The host has started recording and transcription for this meeting. By staying in the meeting, you give consent to being recorded and transcribed.",
```

cwarnermm reviewed (commented):
Provided string feedback inline.

cwarnermm reviewed (commented):
Provided string feedback inline.

matthewbirtch commented:
> @matthewbirtch I implemented some basic heuristics to compact the produced text file output in order to avoid very short sentences. I tried a couple of approaches and eventually converged on a duration based one (as opposed to length) since I felt that splitting/joining based on time gives better results overall as it accounts for pace and potential pauses in speech.
> 
> It's still experimental and highly configurable, right now enforcing the following rules. We join segments if:
> 
> 1. The speaker doesn't change. This is required to guarantee order of the segments (e.g. question/answer sequences).
> 2. There are less than X seconds of pause between the end of a previous text segment and the start of the next one. (X = 2s)
> 3. The overall duration of the current sentence is less than Y seconds. (Y = 10s)
> 
> Give it a try when you get a chance and let me know if you see any improvements or hit any issue.

Definitely seeing an improvement here @streamer45 and I like the rules you've established here - this is really smart! I vote we proceed with what you have and we can always tweak these numbers if we find we're not getting expected results in broader testing.

cwarnermm reviewed (approved):
Thanks for making it so easy for me to review the latest strings with direct links!

streamer45 commented:
Merging this to unblock other PRs. Still working on docs updates as a separate item (https://mattermost.atlassian.net/browse/MM-55587)
